File: app-go/workflows/book_workflow.go
package workflows

import (
	"fmt"
	"temporal-sagas/activities"
	"time"

	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

type BookVacationInput struct {
	BookUserId   string `json:"book_user_id"`
	BookCarId    string `json:"book_car_id"`
	BookHotelId  string `json:"book_hotel_id"`
	BookFlightId string `json:"book_flight_id"`
	Attempts     int    `json:"attempts"`
}

func BookWorkflow(ctx workflow.Context, input BookVacationInput) (string, error) {
	logger := workflow.GetLogger(ctx)
	logger.Info("Book workflow started")

	activityOptions := workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval:    1 * time.Second,
			BackoffCoefficient: 2.0,
			MaximumInterval:    30 * time.Second,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, activityOptions)

	var car string
	err := workflow.ExecuteActivity(ctx, activities.BookCar, input.BookCarId).Get(ctx, &car)
	if err != nil {
		return "", err
	}

	logger.Info("Sleeping for 1 second...")
	workflow.Sleep(ctx, 1*time.Second)

	var hotel string
	err = workflow.ExecuteActivity(ctx, activities.BookHotel, input.BookHotelId).Get(ctx, &hotel)
	if err != nil {
		return "", err
	}

	var flight string
	err = workflow.ExecuteActivity(ctx, activities.BookFlight, input.BookFlightId).Get(ctx, &flight)
	if err != nil {
		return "", err
	}

	output := fmt.Sprintf("%s %s %s", car, hotel, flight)
	return output, nil
}

File: app-python/book_workflow.py
from datetime import timedelta

from temporalio import workflow
from temporalio.common import RetryPolicy

with workflow.unsafe.imports_passed_through():
    from activities import BookVacationInput, book_car, book_flight, book_hotel


@workflow.defn
class BookWorkflow:
    @workflow.run
    async def run(self, input: BookVacationInput):
        compensations = []

        activity_args = {
            "start_to_close_timeout": timedelta(seconds=10),
            "retry_policy": RetryPolicy(
                non_retryable_error_types=["Exception"],
            ),
        }

        try:
            compensations.append("undo_book_car")
            car = await workflow.execute_activity(
                book_car,
                input,
                **activity_args,
            )
            compensations.append("undo_book_hotel")
            hotel = await workflow.execute_activity(
                book_hotel,
                input,
                **activity_args,
            )

            compensations.append("undo_book_flight")
            flight = await workflow.execute_activity(
                book_flight,
                input,
                start_to_close_timeout=timedelta(seconds=10),
                retry_policy=RetryPolicy(
                    initial_interval=timedelta(seconds=1),
                    maximum_interval=timedelta(seconds=1),
                    maximum_attempts=input.attempts,
                    non_retryable_error_types=["Exception"],
                ),
            )
            return car + " " + hotel + " " + flight
        except Exception:
            for compensation in reversed(compensations):
                await workflow.execute_activity(
                    compensation,
                    input,
                    start_to_close_timeout=timedelta(seconds=10),
                )

            return "Booking cancelled"
